vm/src/vm.rs

```diff
@@ -52,13 +53,28 @@ impl VirtualMachine {
             sys_module: sysmod,
             stdlib_inits,
             ctx: ctx,
-            current_frame: None,
+            frames: vec![],
         }
     }
 
     pub fn run_code_obj(&mut self, code: PyObjectRef, scope: PyObjectRef) -> PyResult {
-        let mut frame = Frame::new(code, scope);
-        frame.run_frame_full(self)
+        self.run_frame_full(Frame::new(code, scope))
+    }
+
+    pub fn run_frame_full(&mut self, frame: Frame) -> PyResult {
+        match self.run_frame(frame)? {
+            ExecutionResult::Return(value) => Ok(value),
+            _ => panic!("Got unexpected result from function"),
+        }
+    }
+
+    pub fn run_frame(&mut self, frame: Frame) -> Result<ExecutionResult, PyObjectRef> {
+        let frame = self.ctx.new_frame(frame);
+        self.frames.push(frame.clone());
+        let mut frame = objframe::get_value(&frame);
```

Comment:
Maybe we can seperate the frame into a part that is immutable (like the codeobject) and a part that is mutable (like the current program counter `lasti`). Or at least give the frame a PyRef to the codeobject.